lib/strict/methods/verifiable_method.rb

```diff
@@ -0,0 +1,158 @@
+# frozen_string_literal: true
+
+require "forwardable"
+
+module Strict
+  module Methods
+    class VerifiableMethod # rubocop:disable Metrics/ClassLength
+      extend Forwardable
+
+      class UnknownParameterError < Error
+        attr_reader :parameter_name
+
+        def initialize(parameter_name:)
+          super(message_from(parameter_name:))
+
+          @parameter_name = parameter_name
+        end
+
+        private
+
+        def message_from(parameter_name:)
+          "Strict tried to find a parameter named #{parameter_name} but was unable. " \
+            "It's likely this in an internal bug, feel free to open an issue at #{Strict::ISSUE_TRACKER} for help."
+        end
+      end
+
+      def_delegator :method, :name
+
+      attr_reader :parameters, :returns
+
+      def initialize(method:, parameters:, returns:, instance:)
+        @method = method
+        @parameters = parameters
+        @parameters_index = parameters.to_h { |p| [p.name, p] }
+        @returns = returns
+        @instance = instance
+      end
+
+      def to_s
+        "#{method.owner}#{separator}#{name}"
+      end
+
+      def verify_definition!
+        expected_parameters = Set.new(parameters.map(&:name))
+        defined_parameters = Set.new(method.parameters.filter_map { |kind, name| name unless kind == :block })
+        return if expected_parameters == defined_parameters
+
+        missing_parameters = expected_parameters - defined_parameters
+        additional_parameters = defined_parameters - expected_parameters
+        raise Strict::MethodDefinitionError.new(verifiable_method: self, missing_parameters:, additional_parameters:)
+      end
+
+      # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity, Metrics/BlockLength
+      # TODO(kkt): clean this up- it's late, though, and the tests are passing
+      def verify_parameters!(*args, **kwargs)
```

Comment:
Oof this one was a lot. The tests are passing, but I'm going to circle back on this. There's a lot to track between the validation issues and the re-mapped arguments (from defaulting/coercion). It might make sense to introduce some kind of stateful object to track it (something like an `Arguments` object)